src/tests/test_server_stream.cpp

```diff
@@ -0,0 +1,520 @@
+         }
+
+      void stop_timer()
+         {
+         m_timer.cancel();
+         }
+
+      void check_ec(const std::string& msg, const error_code& ec)
+         {
+         if(ec)
+            { m_result.test_failure(msg, ec.message()); }
+         else
+            { m_result.test_success(msg); }
+         }
+
+      void confirm(const std::string& msg, bool condition)
+         {
+         m_result.confirm(msg, condition);
+         }
+
+   private:
+      net::system_timer m_timer;
+      Result m_result;
+   };
+
+class Server : public Side, public std::enable_shared_from_this<Server>
+   {
+   public:
+      Server(net::io_context& ioc)
+         : Side(server_cert(), server_key()),
+           m_acceptor(ioc),
+           m_result(ioc, "Server"),
+           m_short_read_expected(false) {}
+
+      void listen()
+         {
+         error_code ec;
+         const auto endpoint = k_endpoints.back();
+
+         m_acceptor.open(endpoint.protocol(), ec);
+         m_acceptor.set_option(net::socket_base::reuse_address(true), ec);
+         m_acceptor.bind(endpoint, ec);
+         m_acceptor.listen(net::socket_base::max_listen_connections, ec);
+
+         m_result.check_ec("listen", ec);
+
+         m_result.set_timer("accept");
+         m_acceptor.async_accept(std::bind(&Server::start_session, shared_from_this(), _1, _2));
+         }
+
+      void expect_short_read()
+         {
+         m_short_read_expected = true;
+         }
+
+      void shutdown()
+         {
+         m_result.set_timer("shutdown");
+         m_stream->async_shutdown(std::bind(&Server::on_shutdown, shared_from_this(), _1));
+         }
+
+      Result result() { return m_result.result(); }
+
+   private:
+      void start_session(const error_code& ec, tcp::socket socket)
+         {
+         // Note: If this fails with 'Operation canceled', it likely means the timer expired and the port is taken.
+         m_result.check_ec("accept", ec);
+
+         // Note: If this was a real server, we should create a new session (with its own stream) for each accepted
+         // connection. In this test we only have one connection.
+         m_stream = std::unique_ptr<ssl_stream>(new ssl_stream(std::move(socket), m_ctx));
+
+         m_result.set_timer("handshake");
+         m_stream->async_handshake(Botan::TLS::Connection_Side::SERVER,
+                                   std::bind(&Server::handshake, shared_from_this(), _1));
+         }
+
+      void handshake(const error_code& ec)
+         {
+         m_result.check_ec("handshake", ec);
+         handle_write(error_code{});
+         }
+
+      void handle_write(const error_code& ec)
+         {
+         m_result.check_ec("send_response", ec);
+         m_result.set_timer("read_message");
+         net::async_read(*m_stream, buffer(),
+                         std::bind(&Server::handle_read, shared_from_this(), _1, _2));
+         }
+
+      void handle_read(const error_code& ec, size_t bytes_transferred=0)
```

Comment:
As long as no error occurs, the server will simply echo what it read, looping between `handle_write` and `handle_read`.